vulkan: Add State Space Model (SSM) Operations Support

giuseppe opened the pull request:
implement SSM scan and SSM conv for Vulkan.

Intel Arc:

ggml_vulkan: 0 = Intel(R) Arc(tm) Graphics (MTL) (Intel open-source Mesa driver) | uma: 1 | fp16: 1 | bf16: 0 | warp size: 32 | shared memory: 65536 | int dot: 1 | matrix cores: none

| model                          |       size |     params | backend    | ngl | dev          |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------------ | --------------: | -------------------: |
| granitehybrid ?B Q8_0          |   7.73 GiB |     6.94 B | Vulkan     |  99 | Vulkan0      |           pp512 |          5.27 ± 0.00 |
| granitehybrid ?B Q8_0          |   7.73 GiB |     6.94 B | Vulkan     |  99 | Vulkan0      |           tg128 |          2.33 ± 0.00 |

build: bc07349a7 (6756)

ggml_vulkan: 0 = Intel(R) Arc(tm) Graphics (MTL) (Intel open-source Mesa driver) | uma: 1 | fp16: 1 | bf16: 0 | warp size: 32 | shared memory: 65536 | int dot: 1 | matrix cores: none

| model                          |       size |     params | backend    | ngl | dev          |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------------ | --------------: | -------------------: |
| granitehybrid ?B Q8_0          |   7.73 GiB |     6.94 B | Vulkan     |  99 | Vulkan0      |           pp512 |        213.36 ± 0.00 |
| granitehybrid ?B Q8_0          |   7.73 GiB |     6.94 B | Vulkan     |  99 | Vulkan0      |           tg128 |         15.46 ± 0.00 |

build: a4d94598e (6769)

NVIDIA:

ggml_vulkan: 0 = NVIDIA L40S (NVIDIA) | uma: 0 | fp16: 1 | bf16: 0 | warp size: 32 | shared memory: 49152 | int dot: 0 | matrix cores: KHR_coopmat

| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| granitehybrid ?B Q8_0          |   7.73 GiB |     6.94 B | Vulkan     |  99 |           pp512 |       2042.60 ± 5.26 |
| granitehybrid ?B Q8_0          |   7.73 GiB |     6.94 B | Vulkan     |  99 |           tg128 |         39.57 ± 0.05 |

build: 554fd578a (6766)

ggml_vulkan: 0 = NVIDIA L40S (NVIDIA) | uma: 0 | fp16: 1 | bf16: 0 | warp size: 32 | shared memory: 49152 | int dot: 0 | matrix cores: KHR_coopmat

| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| granitehybrid ?B Q8_0          |   7.73 GiB |     6.94 B | Vulkan     |  99 |           pp512 |      5736.47 ± 24.70 |
| granitehybrid ?B Q8_0          |   7.73 GiB |     6.94 B | Vulkan     |  99 |           tg128 |        140.11 ± 0.32 |

build: a4d94598e (6769)

jeffbolznv reviewed (changes requested):
Thanks for this contribution!

giuseppe commented on ggml/src/ggml-vulkan/vulkan-shaders/ssm_conv.comp:
```diff
@@ -0,0 +1,39 @@
+
+#include "types.glsl"
+
+layout(local_size_x = 1, local_size_y = 1, local_size_z = 1) in;
```
the Intel Arc integrated GPU seems to hang more often if I set it higher.   Since it occasionally hangs even with one thread, I'll investigate that later

0cc4m replied:
Which ARC GPU? In any case, a GPU hang is either a driver bug or a misuse of the Vulkan API (that should be visible if you check with validation layers enabled), it should not be caused by any (valid) workgroup size. A workgroup size of 1 is very inefficient, it would leave all other threads in a warp/wave unused.

giuseppe commented:
I've addressed the comments and pushed a new version.  The results are even better now:

```
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = NVIDIA L4 (NVIDIA) | uma: 0 | fp16: 1 | bf16: 0 | warp size: 32 | shared memory: 49152 | int dot: 0 | matrix cores: KHR_coopmat
| model                          |       size |     params | backend    | ngl | threads | n_batch | n_ubatch | mmap |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ------: | ------: | -------: | ---: | --------------: | -------------------: |
| granitehybrid ?B Q8_0          |   7.73 GiB |     6.94 B | Vulkan     |  99 |       4 |     512 |     4096 |    0 |           pp512 |      2569.27 ± 19.37 |
| granitehybrid ?B Q8_0          |   7.73 GiB |     6.94 B | Vulkan     |  99 |       4 |     512 |     4096 |    0 |           tg128 |         80.77 ± 0.12 |
```

0cc4m commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -9027,6 +9067,177 @@ static void ggml_vk_rwkv_wkv7(ggml_backend_vk_context * ctx, vk_context& subctx,
+    }, pc, elements);
+}
+
+static void ggml_vk_ssm_conv(ggml_backend_vk_context * ctx, vk_context& subctx, ggml_tensor * dst, bool dryrun = false) {
```
Is there a reason why this isn't just calling `ggml_vk_op_f32`? With 2 input sources, that should be fine.

giuseppe commented on ggml/src/ggml-vulkan/vulkan-shaders/ssm_scan.comp:
```diff
@@ -0,0 +1,153 @@
+    warp_sdata[warp_offset + lane] = val;
+    barrier();
+
+    if (lane < 16) warp_sdata[warp_offset + lane] += warp_sdata[warp_offset + lane + 16];
```
sorry I've missed this one.  Yeah I don't think it would work with a size != 32.  I need to think more through this one.

Do you've any suggestions on what I could do here?

jeffbolznv replied:
You could change this to a while loop that will handle any power-of-two value of WARP_SIZE. We do want to allow the spec constants to be changeable but it's fine to have limitations like "must be a power of two".

jeffbolznv commented on ggml/src/ggml-vulkan/vulkan-shaders/vulkan-shaders-gen.cpp:
```diff
@@ -919,6 +919,12 @@ void process_shaders() {
 
+    string_to_spv("ssm_scan_f32_d16", "ssm_scan.comp", {{"A_TYPE", "float"}});
+    string_to_spv("ssm_scan_f32_d128", "ssm_scan.comp", {{"A_TYPE", "float"}});
+    string_to_spv("ssm_scan_f32_d256", "ssm_scan.comp", {{"A_TYPE", "float"}});
```
These three are all identical now, you only need one.

giuseppe committed:
vulkan: implement SSM scan operation

Add State Space Model scan operation to the Vulkan backend.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

giuseppe committed:
vulkan: implement SSM conv operation

Add State Space Model conv operation to the Vulkan backend.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

0cc4m reviewed (commented):
It works fine now.

jeffbolznv commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -8592,6 +8628,14 @@ static void ggml_vk_op_f32(ggml_backend_vk_context * ctx, vk_context& subctx, co
+            const uint32_t nr  = src0->ne[1];
+            const uint32_t n_t = dst->ne[1];
+            const uint32_t n_s = dst->ne[2];
+            elements = { CEIL_DIV(nr, 32), n_t, n_s };
```
Seems like you could use wg_denoms for this one (make wg_denoms = {32, 1, 1} and not explicitly use CEIL_DIV here).

jeffbolznv commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -12879,6 +13048,57 @@ static bool ggml_backend_vk_device_supports_op(ggml_backend_dev_t dev, const ggm
+                if (ggml_is_quantized(op->src[0]->type) || ggml_is_quantized(op->src[1]->type) || ggml_is_quantized(op->src[2]->type)) {
+                    return false;
+                }
+                if (op->src[3] && ggml_is_quantized(op->src[3]->type)) {
```
You could maybe do these ggml_is_quantized checks in a loop or using any_of.

jeffbolznv reviewed (approved):
LGTM. All my comments have been addressed. I haven't had a chance to actually run it yet, I can try to do that tomorrow, but I don't want to block the change.

jeffbolznv commented:
Oh, there's a failure in the lavapipe CI:

```
2025-10-16T04:27:04.7561155Z 31: [SSM_SCAN] NMSE = 0.581159257 > 0.000000100   SSM_SCAN(type=f32,d_state=256,head_dim=64,n_head=8,n_group=2,n_seq_tokens=32,n_seqs=4): [1;31mFAIL[0m
```

lavapipe uses an unusual subgroup size, that might be related.

giuseppe commented:
> > Oh, there's a failure in the lavapipe CI:
> > ```
> > 2025-10-16T04:27:04.7561155Z 31: [SSM_SCAN] NMSE = 0.581159257 > 0.000000100   SSM_SCAN(type=f32,d_state=256,head_dim=64,n_head=8,n_group=2,n_seq_tokens=32,n_seqs=4): �[1;31mFAIL�[0m
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > lavapipe uses an unusual subgroup size, that might be related.
> 
> how do I run this locally? In what CI task is it happening?

nevermind, I see it now.  Taking a look

giuseppe commented:
to simplify the review, I post here the patch I've applied on the top of the previous version:

```diff
diff --git a/ggml/src/ggml-vulkan/vulkan-shaders/ssm_scan.comp b/ggml/src/ggml-vulkan/vulkan-shaders/ssm_scan.comp
index 06aa10bfe..daebb5dc0 100644
--- a/ggml/src/ggml-vulkan/vulkan-shaders/ssm_scan.comp
+++ b/ggml/src/ggml-vulkan/vulkan-shaders/ssm_scan.comp
@@ -96,7 +96,7 @@ void main() {
             barrier();
         }

-        [[unroll]] for (uint j = 0; j < SPLIT_H / (D_STATE / SUBGROUP_SIZE); j++) {
+        [[unroll]] for (uint j = 0; j <= SPLIT_H / (D_STATE / SUBGROUP_SIZE); j++) {
             const uint idx = (tid % SUBGROUP_SIZE) +
                             D_STATE * (tid / SUBGROUP_SIZE) +
                             j * D_STATE * (D_STATE / SUBGROUP_SIZE);
@@ -104,13 +104,13 @@ void main() {
             uint lane = tid % SUBGROUP_SIZE;

             [[unroll]] for (uint offset = SUBGROUP_SIZE / 2; offset > 0; offset >>= 1) {
-                if (lane < offset) {
+                if (idx < SPLIT_H * D_STATE) {
                     stateC[idx] += stateC[idx + offset];
                 }
                 barrier();
             }

-            if (tid % SUBGROUP_SIZE == 0) {
+            if (idx < SPLIT_H * D_STATE && tid % SUBGROUP_SIZE == 0) {
                 const uint k = tid / SUBGROUP_SIZE + j * (D_STATE / SUBGROUP_SIZE);
                 d[y_base_idx + i * stride_y + k] = stateC[idx];
             }
```

the test passes locally now

jeffbolznv commented:
Passes for me, too.

freebsd-nils-level1 commented:
Wow, that really is a game changer for Vulkan on nvidia cards. Here an example of my 4060 Mobile GPU:
.
Original:
```
·• env GGML_VK_VISIBLE_DEVICES=1 .build-vulkan/bin/llama-bench -m ~/work/.models/Huihui-granite-4.0-h-tiny-abliterated.Q8_0.gguf 
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = NVIDIA GeForce RTX 4060 Laptop GPU (NVIDIA) | uma: 0 | fp16: 1 | bf16: 1 | warp size: 32 | shared memory: 49152 | int dot: 1 | matrix cores: NV_coopmat2
| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| granitehybrid ?B Q8_0          |   6.88 GiB |     6.94 B | Vulkan     |  99 |           pp512 |      1277.04 ± 51.92 |
| granitehybrid ?B Q8_0          |   6.88 GiB |     6.94 B | Vulkan     |  99 |           tg128 |         28.11 ± 1.00 |

build: 683fa6ba4 (6781)

··• env GGML_VK_VISIBLE_DEVICES=1 .build-cublas/bin/llama-bench -m ~/work/.models/Huihui-granite-4.0-h-tiny-abliterated.Q8_0.gguf 
ggml_cuda_init: GGML_CUDA_FORCE_MMQ:    no
ggml_cuda_init: GGML_CUDA_FORCE_CUBLAS: no
ggml_cuda_init: found 1 CUDA devices:
  Device 0: NVIDIA GeForce RTX 4060 Laptop GPU, compute capability 8.9, VMM: yes
| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| granitehybrid ?B Q8_0          |   6.88 GiB |     6.94 B | CUDA       |  99 |           pp512 |      3358.79 ± 16.26 |
| granitehybrid ?B Q8_0          |   6.88 GiB |     6.94 B | CUDA       |  99 |           tg128 |         85.83 ± 0.61 |

build: 683fa6ba4 (6781)
```
.
Modified for SSM support:
```
··• env GGML_VK_VISIBLE_DEVICES=1 bin/llama-bench -m ~work/.models/Huihui-granite-4.0-h-tiny-abliterated.Q8_0.gguf
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = NVIDIA GeForce RTX 4060 Laptop GPU (NVIDIA) | uma: 0 | fp16: 1 | bf16: 1 | warp size: 32 | shared memory: 49152 | int dot: 1 | matrix cores: NV_coopmat2
| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| granitehybrid ?B Q8_0          |   6.88 GiB |     6.94 B | Vulkan     |  99 |           pp512 |      2410.88 ± 26.99 |
| granitehybrid ?B Q8_0          |   6.88 GiB |     6.94 B | Vulkan     |  99 |           tg128 |         93.11 ± 0.68 |

build: c3313203 (6776)
```
.
It really gets close to Cuda; even surpasses its text generation speed. Prompt processing still lacks a bit behind.
.
.
And it even helps on an AMD 780M GPU - although not that much for prompt processing unfortunately.

Original:
```
··• env GGML_VK_VISIBLE_DEVICES=0 .build-vulkan/bin/llama-bench -m ~/work/.models/Huihui-granite-4.0-h-tiny-abliterated.Q8_0.gguf 
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = AMD Radeon 780M Graphics (RADV PHOENIX) (radv) | uma: 1 | fp16: 1 | bf16: 0 | warp size: 64 | shared memory: 65536 | int dot: 1 | matrix cores: KHR_coopmat
| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| granitehybrid ?B Q8_0          |   6.88 GiB |     6.94 B | Vulkan     |  99 |           pp512 |        472.07 ± 3.40 |
| granitehybrid ?B Q8_0          |   6.88 GiB |     6.94 B | Vulkan     |  99 |           tg128 |         21.56 ± 0.40 |

build: 683fa6ba4 (6781)
```
.
Modified:
```
··• env GGML_VK_VISIBLE_DEVICES=0 bin/llama-bench -m ~/work/.models/Huihui-granite-4.0-h-tiny-abliterated.Q8_0.gguf
ggml_vulkan: Found 1 Vulkan devices:
ggml_vulkan: 0 = AMD Radeon 780M Graphics (RADV PHOENIX) (radv) | uma: 1 | fp16: 1 | bf16: 0 | warp size: 64 | shared memory: 65536 | int dot: 1 | matrix cores: KHR_coopmat
| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| granitehybrid ?B Q8_0          |   6.88 GiB |     6.94 B | Vulkan     |  99 |           pp512 |        513.78 ± 6.68 |
| granitehybrid ?B Q8_0          |   6.88 GiB |     6.94 B | Vulkan     |  99 |           tg128 |         33.78 ± 0.21 |

build: c3313203 (6776)
```

I understand that one needs special SSM/Mamba(?) LLM models in order to get this. Does anyone know how to specifically search for these kind of models on Huggingface?


Thanks so much for your work on this, giuseppe, and please get this into upstream...

Regards,
Nils

0cc4m commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -12879,6 +13048,50 @@ static bool ggml_backend_vk_device_supports_op(ggml_backend_dev_t dev, const ggm
+                const uint32_t MAX_D_STATE = 256;
+
+                size_t stateC_size = SPLIT_H * MAX_D_STATE * sizeof(float);
+                size_t warp_sdata_size = MAX_D_STATE * sizeof(float);
```
You forgot to update this calculation.

0cc4m commented on ggml/src/ggml-vulkan/ggml-vulkan.cpp:
```diff
@@ -12879,6 +13048,50 @@ static bool ggml_backend_vk_device_supports_op(ggml_backend_dev_t dev, const ggm
+                const uint32_t SPLIT_H = 16;
+                const uint32_t MAX_D_STATE = 256;
+
+                size_t stateC_size = SPLIT_H * MAX_D_STATE * sizeof(float);
```
Shouldn't this be using d_state instead of MAX_D_STATE, since the smaller shader may fit even if the large one does not?

0cc4m reviewed (commented):
It works as intended, I just have some comments about the supports_op code.

giuseppe commented:
@0cc4m fixed the last comments and also added the new operations to `docs/ops.md`, that I forgot to do earlier

0cc4m reviewed (approved):
Thank you, looks good now.
